DOCS-3916: Add more DoCommand explanation (#4259)
viamrobotics/docs: docs/dev/reference/sdks/docommand.md

---
title: "Implement or use DoCommand"
linkTitle: "DoCommand"
weight: 67
type: "docs"
description: "Implement DoCommand in your module or use it from Viam's SDKs."
tags: ["sdk", "extend"]
date: "2025-04-29"
# updated: ""  # When the content was last entirely checked
---

The `DoCommand` method is a generic wrapper that you can use to send commands that don't fit into any other method.

> `DoCommand` is part of every [component](/dev/reference/apis/#component-apis) and [service](/dev/reference/apis/#service-apis), though most models do not implement it.

Author: ```suggestion
`DoCommand` is part of every [component](/dev/reference/apis/#component-apis) and [service API](/dev/reference/apis/#service-apis), though most models do not implement it.
```

---
title: "Implement or use DoCommand"
linkTitle: "DoCommand"
weight: 67
type: "docs"
description: "Implement DoCommand in your module or use it from Viam's SDKs."
tags: ["sdk", "extend"]
date: "2025-04-29"
# updated: ""  # When the content was last entirely checked
---

> The `DoCommand` method is a flexible wrapper that you can use to send commands that don't fit into any other method.

Reviewer: ```suggestion
The `DoCommand` method is a flexible wrapper that you can use to send commands that have no corresponding built-in API method.
```
Maybe something like this? I don't like the "don't fit into any other method" because i think it's odd to think of commands as inside methods

`DoCommand` is part of every [component](/dev/reference/apis/#component-apis) and [service API](/dev/reference/apis/#service-apis), though most models do not implement it.

In the majority of cases, you should use a more specific method for your component or service.
As the developer of a resource, you can implement `DoCommand` in your module if you need to add custom commands to your resource.

As the user of a resource, you can only call `DoCommand` if it is implemented in the model you are using.

> Refer to the model's documentation (for example, the module README) to see whether `DoCommand` is implemented and how to use it.

Reviewer: ```suggestion
Refer to the model's documentation on the module page to see whether `DoCommand` is implemented and how to use it.
```
All docs should now be on the module pages for supporter models

Author: Most modules don't actually have it pulled in though. Viam official components will when Rand's team is done with that, but most existing ones won't. Over time the ratio will shift as new modules will have this by default, but I think it makes sense to document as-is and change it at such time as it's different.
